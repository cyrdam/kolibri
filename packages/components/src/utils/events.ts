// TODO: Should be synchronized with enums/events.ts
export type KoliBriEventType =
	| 'blur'
	| 'change'
	| 'changePage'
	| 'changePageSize'
	| 'click'
	| 'close'
	| 'focus'
	| 'input'
	| 'mousedown'
	| 'reset'
	| 'selection-change' // todo remove
	| 'selectionChange'
	| 'sort'
	| 'submit'
	| 'toggle';

export function stopPropagation(event: Event): void {
	event.stopImmediatePropagation();
	event.stopPropagation();
}

function createKoliBriEvent<T>(type: KoliBriEventType, detail: T): CustomEvent {
	const event = new CustomEvent(`kol-${type}`, {
		bubbles: true,
		cancelable: true,
		composed: true,
		detail: detail,
	});
	return event;
}

function dispatchKoliBriEvent<T>(target: EventTarget, type: KoliBriEventType, detail?: T): boolean {
	const dispatch = target.dispatchEvent(createKoliBriEvent(type, detail));
	return dispatch;
}

/**
 * @deprecated
 */
export function tryToDispatchKoliBriEvent<T>(type: KoliBriEventType, target?: EventTarget, detail?: T): void {
	target && dispatchKoliBriEvent(target, type, detail);
}

export function dispatchDomEvent(target: HTMLElement, type: KoliBriEventType) {
	target.dispatchEvent(new Event(type, { bubbles: true, composed: true }));
}

export function preventDefaultAndStopPropagation(event: Event) {
	event.preventDefault();
	stopPropagation(event);
}
