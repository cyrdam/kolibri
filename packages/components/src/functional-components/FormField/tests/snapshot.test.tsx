import { h } from '@stencil/core';
import { renderFunctionalComponentToSpecPage } from '../../../utils/testing';
import KolFormFieldFc from '../FormField';
import type { InternMsgPropType } from '../../../schema';

describe('KolFormFieldFc', () => {
	it('should render correctly', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.tagName).toBe('DIV');
	});

	it('should render correctly with fieldset tag', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" component="fieldset" />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.tagName).toBe('FIELDSET');
	});

	it('should render correctly with div tag', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" component="div" />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.tagName).toBe('DIV');
	});

	it('should render with required props', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" required />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.className).toContain('required');
	});

	it('should render with error message', async () => {
		const msg: InternMsgPropType = { type: 'error', description: 'Error message' };
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" msg={msg} touched />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.textContent).toContain('Error message');
	});

	it('should render with hint', async () => {
		const hint = 'Hint message';
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" hint={hint} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.textContent).toContain(hint);
	});

	it('should render with counter', async () => {
		const counter = { currentLength: 5, maxLength: 10 };
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" counter={counter} />);
		expect(page.root).toMatchSnapshot();
		expect(page.root?.textContent).toContain(5);
		expect(page.root?.textContent).toContain(10);
	});

	it('should render with tooltip', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => <KolFormFieldFc id="test-id" label="Test Label" tooltipAlign="right" />);
		expect(page.root).toMatchSnapshot();
	});

	it('should render with custom children', async () => {
		const page = await renderFunctionalComponentToSpecPage(() => (
			<KolFormFieldFc id="test-id" label="Test Label">
				<input type="text" />
			</KolFormFieldFc>
		));
		expect(page.root).toMatchSnapshot();
		expect(page.root?.getElementsByTagName('input')).toHaveLength(1);
	});
});
