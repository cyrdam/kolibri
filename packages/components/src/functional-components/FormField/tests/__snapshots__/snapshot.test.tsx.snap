// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KolFormFieldFc should render correctly 1`] = `
<div class="kol-form-field" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
</div>
`;

exports[`KolFormFieldFc should render correctly with div tag 1`] = `
<div class="kol-form-field" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
</div>
`;

exports[`KolFormFieldFc should render correctly with fieldset tag 1`] = `
<fieldset class="kol-form-field" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
</fieldset>
`;

exports[`KolFormFieldFc should render with counter 1`] = `
<div class="kol-form-field" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
  <span aria-atomic="true" aria-live="polite" class="kol-form-field__counter" data-testid="input-counter">
    5
    <span aria-label="kol-of" role="img">
      /
    </span>
    10
    <span>
      kol-characters
    </span>
  </span>
</div>
`;

exports[`KolFormFieldFc should render with custom children 1`] = `
<div class="kol-form-field" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input">
    <input type="text">
  </div>
</div>
`;

exports[`KolFormFieldFc should render with error message 1`] = `
<div class="kol-form-field kol-form-field--error kol-form-field--msg-type-error kol-form-field--touched" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
  <div aria-hidden="true" class="kol-alert kol-alert--error kol-alert--msg kol-form-field__msg" description="Error message" id="test-id-msg">
    <div class="kol-alert__container">
      <kol-icon _icons="codicon codicon-error" _label="kol-error" class="kol-alert__heading-icon"></kol-icon>
      <div class="kol-alert__container-content">
        <div class="kol-alert__content">
          Error message
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`KolFormFieldFc should render with hint 1`] = `
<div class="kol-form-field" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
  <span class="kol-form-field__hint" id="test-id-hint">
    Hint message
  </span>
</div>
`;

exports[`KolFormFieldFc should render with required props 1`] = `
<div class="kol-form-field kol-form-field--required" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
</div>
`;

exports[`KolFormFieldFc should render with tooltip 1`] = `
<div class="kol-form-field" role="presentation">
  <label class="kol-form-field__label" htmlfor="test-id" id="test-id-label">
    <span class="kol-form-field__label-text">
      <span>
        Test Label
      </span>
    </span>
  </label>
  <div class="kol-form-field__input"></div>
</div>
`;
